### PR TITLE
Introduce 'cancancan' gem for authorization

### PR DESCRIPTION
**Pull request by ericfreese**
## Overview
There's a lot of duplicated logic for permissions-checking scattered across the app. [CanCan](https://github.com/CanCanCommunity/cancancan) will help centralize our permissions logic.

## Notes
- This is a first step. Future PRs can replace more permissions code with `authorize!` or other `CanCan` methods.
- This PR also standardizes the behavior of the application when a user tries to do something that they're not authorized for to always display a flash message and redirect back (or to `root_path` if no referrer is available).

### CODE REVIEW

File: app/controllers/regions_controller.rb [hunk 1]
```diff
@@ -47,7 +43,9 @@ def update
   end
 
   def destroy
-    Region.find(params[:id]).destroy
+    region = Region.find(params[:id])
+    authorize! :destroy, @region
```
Comment (Fosome) [inline app/controllers/regions_controller.rb]: `region` vs `@region`

File: app/models/ability.rb [hunk 2]
```diff
@@ -18,4 +19,27 @@ def super_admin?
...
+  def assignments
+    @assignments ||= volunteer.assignments
+  end
+
+  def admin_assignments
+    @admin_assignments ||= assignments.where(admin: true)
+  end
+
+  def admin_region_ids
+    admin_assignments.map(&:region_id)
```
Comment (Fosome) [inline app/models/ability.rb]: Can we use SQL here too with `pluck`?

File: spec/models/ability_spec.rb [hunk 3]
```diff
@@ -10,4 +10,65 @@
...
+      end
+
+      it 'cannot update other regions' do
+        other_region = create(:region)
+        expect(subject.can?(:update, other_region)).to eq(false)
+      end
+    end
+
+    describe 'food types' do
+      let(:food_type) { create(:food_type, region: region) }
```
Comment (Fosome) [inline spec/models/ability_spec.rb]: You may be able to use `build` on a lot of these resources.  If actually persisting the thing isn't required - can speed up this test.

Non-blocking

**Comment by rylanb** [comment]
Out of curiousity, why CanCan/CanCanCan vs. Pundit?

**Comment by ericfreese** [comment]
@rylanb Really just because I'm a little more familiar with CanCan. What are your thoughts on pros/cons of the two? I'd be happy to do some exploration of implementing pundit instead.

**Comment by rylanb** [comment]
@ericfreese I was a fan of CanCanCan, too. I read some comparisons and it seemed to indicate that Pundit was lighter weight and didn't evaluate the entire class like CCC does on each page load. Pundit uses POROs and fairly easy to use, but don't waste too much time on it, was just curious

https://www.quora.com/What-are-the-key-factors-to-consider-before-choosing-Pundit-or-CanCanCan-for-authorization-in-a-Rails-project

https://resources.distilnetworks.com/all-blog-posts/cancan-vs-pundit-i-choose-pundit-for-authorization

**Comment by ericfreese** [comment]
I'm gonna take a few minutes to implement pundit in a separate PR and we can compare :) Sounds like pundit's probably the way to go.

**Comment by rylanb** [comment]
Sure poke at it, always good to try something new. But again, not top priority, was more just curious what the reasoning was, is all.

**Comment by ericfreese** [comment]
@rylanb I pushed up a WIP branch mostly implementing Pundit: https://github.com/boulder-food-rescue/food-rescue-robot/compare/ef/pundit_auth

First reactions are:

- Pundit has a lot more boilerplate than CanCan (more verbose). Pundit requires implementing permissions as well as scope, whereas CanCan implements the scope based on the permissions.
- Pundit looks to be more flexible than CanCan for implementing complex authorization rules.
- Pundit only loads the permissions you need to check, as opposed to CanCan loading all permissions every time you need to check one of them.
- Pundit breaks permissions down by ActiveRecord model, which makes it more difficult (at least for me) to comprehend all permissions for a role (super admin, region admin, regular volunteer) at once. CanCan seems to provide better mechanism for grouping permissions by role.

My overall thought is that CanCan will be better for this app, given the relatively simple, hierarchical nature of the authorization rules (i.e. Super admins can do everything region admins can do, region admins can do everything regular volunteers can do). CanCan's [ability precedence](https://github.com/CanCanCommunity/cancancan/wiki/Ability-Precedence) seems to be a natural fit, adding on preferences in a cascade.

I think the performance impact of loading all permissions (as CanCan does) will not be felt because the permissions are relatively easy to calculate, typically requiring only one query to fetch the regions that the user is an admin for.

One criticism of CanCan is that the `ability.rb` file can grow unwieldy, but I think we can be aware of that danger and split it up into separate classes/modules/methods as necessary.

All that being said, I really don't have much experience with Pundit, so it could be that I'm totally missing something or am just implementing it in a shitty way ;)

**Comment by rylanb** [comment]
No i think that's a fair review. I think CanCanCan is fine. 👍

**Comment by ericfreese** [comment]
Ok sounds good. I'm on another project at MojoTech now so I'll have quite a bit less time going forward for BFR, but I'll address some of @Fosome's feedback when I have time so we can get this merged!

File: app/controllers/logs_controller.rb [hunk 4]
```diff
@@ -192,6 +180,7 @@ def create
 
   def show
     @log = Log.find(params[:id])
+    authorize! :read, @Log
```
Comment (ericfreese) [inline app/controllers/logs_controller.rb]: fixing...

**Comment by ericfreese** [comment]
Ok @Fosome @rylanb, I think this should be ready for another review.

**Pull request merged**